lib/runtime-c-api/build.rs

```diff
@@ -14,18 +14,21 @@ fn main() {
 
     let mut pre_header = r#"
 #if !defined(WASMER_H_MACROS)
+
 #define WASMER_H_MACROS
 
-#if defined(MSVC)
-#if defined(_M_AMD64)
-#define ARCH_X86_64
-#endif
+// Define the `ARCH_X86_X64` constant.
+#if defined(MSVC) && defined(_M_AMD64)
+#  define ARCH_X86_64
+#elif (defined(GCC) || defined(__GNUC__) || defined(__clang__)) && defined(__x86_64__)
+#  define ARCH_X86_64
 #endif
 
+// Define the `DEPRECATED` macro.
 #if defined(GCC) || defined(__GNUC__) || defined(__clang__)
```

Comment:
While I think this should work as written, I don't think `__clang__` should be listed here. Clang can run in two modes, either emulating GCC which means both defining the `__GNUC__` macro and implementing gcc's `__attribute__` extensions, or emulating MSVC which means both defining `MSVC` and implementing `__declspec`. It turns out that this will work as is because even if clang is running as MSVC-compatible, it still understands `__attribute__` from GCC. However, I think it'd be better to just leave it out.

Or we can add specific clang support like we have specific GCC and MSVC support:
```
#ifndef __has_attribute
  #define __has_attribute(x) 0  // Compatibility with non-clang compilers.
#endif
#ifndef __has_declspec_attribute
  #define __has_declspec_attribute(x) 0  // Compatibility with non-clang compilers.
#endif
```
Those `__has_attribute` and `__has_declspec_attribute` are clang inventions, but designed with the idea that other compilers could implement the same thing. With the code above you could then write `#if defined(GCC) || defined(__GNUC__) || __has_attribute(deprecated)` here and `#elif defined(MSVC) || __has_declspec_attribute(deprecated)` below.

Reply:
Thanks, it makes a lot of sense! I learned something today!